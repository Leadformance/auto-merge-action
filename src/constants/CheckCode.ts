export enum CheckCode {
  invalid = 'INVALID',
  valid = 'VALID',
  ignored = 'IGNORED',
}
