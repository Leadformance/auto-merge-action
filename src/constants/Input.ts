export enum Input {
  token = 'token',
  mergeMethod = 'mergeMethod',
  enableApproval = 'enableApproval',
  enableReadyForReview = 'enableReadyForReview',
  title = 'title',
  labels = 'labels',
  users = 'users',
  teams = 'teams',
}
