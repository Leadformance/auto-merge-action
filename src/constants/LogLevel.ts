export enum LogLevel {
  debug,
  info,
  warning,
  error,
}
