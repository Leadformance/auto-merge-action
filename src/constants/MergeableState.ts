export enum MergeableState {
  mergeable = 'MERGEABLE',
  conflicting = 'CONFLICTING',
  unknown = 'UNKNOWN',
}
