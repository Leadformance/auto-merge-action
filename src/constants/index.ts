export { CheckCode } from './CheckCode';
export { Input } from './Input';
export { LogLevel } from './LogLevel';
export { MergeableState } from './MergeableState';
export { MergeMethod } from './MergeMethod';
export { Output } from './Output';
export { PullRequestReviewEvent } from './PullRequestReviewEvent';
export { PullRequestState } from './PullRequestState';
export { SUPPORTED_EVENTS } from './SupportedEvents';
