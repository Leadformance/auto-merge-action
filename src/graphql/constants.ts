export const GRAPHQL_GITHUB_API_URL = 'https://api.github.com/graphql';
