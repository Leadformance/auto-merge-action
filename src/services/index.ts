export { CheckService } from './Check.service';
export { InputsService } from './Inputs.service';
export { OrganizationsService } from './Organizations.service';
export { PullRequestsService } from './PullRequests.service';
export { RepositoriesService } from './Repositories.service';
