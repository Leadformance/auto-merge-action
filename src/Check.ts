import { ExitCode, info } from '@actions/core';
import { Context } from '@actions/github/lib/context';

import { Inputs } from './@types';
import {
  CheckCode,
  Input,
  MergeableState,
  PullRequestState,
} from './constants';
import { CheckService } from './services';

export class Check {
  private readonly service: CheckService;

  constructor(
    private readonly inputs: Inputs,
    private readonly context: Context
  ) {
    this.service = new CheckService(this.inputs.token);
  }

  public async run(): Promise<ExitCode> {
    const {
      eventName,
      repo: { owner, repo },
      job,
      runNumber,
      runId,
      issue,
    } = this.context;

    // Validate context event
    info('Validate context event...');
    if (this.service.isContextEventValid(eventName) === CheckCode.invalid)
      throw new Error(`This action does not supports "${eventName}" events`);

    // Get pull request
    info('Get pull request info...');
    const pullRequest = await this.service.getPullRequest(
      owner,
      repo,
      issue.number
    );

    if (pullRequest.state !== PullRequestState.open)
      throw new Error('The pull request is not opened');

    const author = pullRequest.author.login;

    // Validate title
    await this.validateRequirement(
      Input.title,
      () =>
        this.service.isTitleRequirementValid(
          pullRequest.title,
          this.inputs.title
        ),
      `The title of the pull request does not match the search: ${this.inputs.title}`
    );

    // Validate labels
    await this.validateRequirement(
      Input.labels,
      () =>
        this.service.isLabelsRequirementValid(
          pullRequest.labels,
          this.inputs.labels
        ),
      `The pull request does not have the required labels: ${this.inputs.labels.join(
        ', '
      )}`
    );

    // Validate users
    await this.validateRequirement(
      Input.users,
      () => this.service.isUsersRequirementValid(author, this.inputs.users),
      `User "${author}" is not authorized`
    );

    // Validate teams
    await this.validateRequirement(
      Input.teams,
      async () =>
        this.service.isTeamsRequirementValid(author, this.inputs.teams),
      `User "${author}" is not part of any teams`
    );

    // Approve pull request
    info('Approve pull request...');
    await this.service.approvePullRequest(
      pullRequest.id,
      pullRequest.checksUrl,
      pullRequest.isDraft,
      job,
      runNumber,
      runId,
      this.inputs.enableReadyForReview
    );

    // Merge pull request
    if (pullRequest.mergeable === MergeableState.conflicting)
      throw new Error(
        'The pull request cannot be merged due to merge conflicts'
      );

    if (pullRequest.mergeable === MergeableState.unknown)
      throw new Error('The pull request cannot be merged');

    try {
      info('Enable auto-merge on pull request...');
      await this.service.enableAutoMergeOnPullRequest(
        pullRequest.id,
        this.inputs.mergeMethod
      );
    } catch (error) {
      info(`Auto-merge request failed: ${error.message}`);
      info('Merging the pull request...');
      await this.service.mergePullRequest(
        pullRequest.id,
        this.inputs.mergeMethod
      );
    }

    return ExitCode.Success;
  }

  private async validateRequirement(
    name: string,
    validate: () => CheckCode | Promise<CheckCode>,
    error: string
  ): Promise<void> {
    info(`Validate ${name}...`);

    const code = await validate();

    switch (code) {
      case CheckCode.invalid:
        info(`  ✘︎ Requirement ${name} is not valid`);
        throw new Error(error);
      case CheckCode.valid:
        info(`  ✔︎ Requirement ${name} is valid`);
        break;
      case CheckCode.ignored:
        info(`  ○ Requirement ${name} was ignored`);
        break;
    }
  }
}
