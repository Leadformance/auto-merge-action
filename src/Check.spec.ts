import * as core from '@actions/core';

import { InputsHelper } from '../test/InputsHelper';
import { IMockContext, MockContext } from '../test/MockContext';
import { MockLogger } from '../test/MockLogger';
import { Check } from './Check';
import { Input } from './constants';
import { InputsService } from './services';

jest.mock('./services/Organizations.service.ts');
jest.mock('./services/PullRequests.service.ts');
jest.mock('./services/Repositories.service.ts');

interface EnvInputs {
  [Input.token]?: string;
  [Input.mergeMethod]?: string;
  [Input.enableApproval]?: string;
  [Input.enableReadyForReview]?: string;
  [Input.title]?: string;
  [Input.labels]?: string;
  [Input.users]?: string;
  [Input.teams]?: string;
}

describe('Check', () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  describe('run', () => {
    const mockLogger = new MockLogger();

    beforeAll(() => {
      const logHandler = (error: string | Error) => {
        mockLogger.log(
          `${typeof error === 'string' ? error : `Error: ${error.message}`}`
        );
      };

      jest.spyOn(core, 'debug').mockImplementation(logHandler);
      jest.spyOn(core, 'info').mockImplementation(logHandler);
      jest.spyOn(core, 'warning').mockImplementation(logHandler);
      jest.spyOn(core, 'error').mockImplementation(logHandler);
    });

    afterEach(() => {
      mockLogger.clear();

      InputsHelper.clearInputs([
        Input.token,
        Input.mergeMethod,
        Input.enableApproval,
        Input.enableReadyForReview,
        Input.title,
        Input.labels,
        Input.users,
        Input.teams,
      ]);
    });

    describe('should throw an Error', () => {
      it.each([
        ['with wrong "event"', {}, { eventName: 'lorem' }],
        ['with wrong "title"', { title: 'lorem' }, {}],
        ['with wrong "labels"', { labels: 'lorem' }, {}],
        ['with wrong "users"', { users: 'lorem' }, {}],
        ['with wrong "teams"', { teams: '@foo/bar' }, {}],
        [
          'with pull request conflicts',
          {},
          { payload: { pull_request: { number: 5 } } },
        ],
        [
          'with pull request not mergeable',
          {},
          { payload: { pull_request: { number: 6 } } },
        ],
        [
          'with pull request closed',
          {},
          { payload: { pull_request: { number: 7 } } },
        ],
      ])('%s', async (_, inputs: EnvInputs, context: Partial<IMockContext>) => {
        InputsHelper.setInputs({ [Input.token]: 'token', ...inputs });

        expect.assertions(2);

        try {
          await new Check(
            InputsService.getInputs(),
            new MockContext(context)
          ).run();
        } catch (error) {
          // eslint-disable-next-line jest/no-conditional-expect,jest/no-try-expect
          expect(error).toEqual(new Error(error.message));
          core.error(`❌  ${error.message}`);
        } finally {
          expect(mockLogger.logs).toMatchSnapshot();
        }
      });
    });

    describe('should run without error', () => {
      it.each([
        ['without inputs', {}],
        [
          'with inputs',
          {
            title: '/pull/i',
            labels: 'foo',
            users: 'author',
            teams: '@org/team',
          },
        ],
      ])('%s', async (_, inputs: EnvInputs) => {
        InputsHelper.setInputs({ [Input.token]: 'token', ...inputs });

        expect.assertions(1);

        try {
          await new Check(InputsService.getInputs(), new MockContext()).run();
          core.info('🎉  The pull request will be merged automatically!');
        } catch {
          // eslint-disable-next-line jest/no-conditional-expect,jest/no-try-expect
          expect(true).toEqual(false);
        } finally {
          expect(mockLogger.logs).toMatchSnapshot();
        }
      });
    });
  });
});
