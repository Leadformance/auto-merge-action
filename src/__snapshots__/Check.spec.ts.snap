// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check run should run without error with inputs 1`] = `
"Validate context event...
Get pull request info...
Validate title...
  ✔︎ Requirement title is valid
Validate labels...
  ✔︎ Requirement labels is valid
Validate users...
  ✔︎ Requirement users is valid
Validate teams...
  ✔︎ Requirement teams is valid
Approve pull request...
Enable auto-merge on pull request...
🎉  The pull request will be merged automatically!"
`;

exports[`Check run should run without error without inputs 1`] = `
"Validate context event...
Get pull request info...
Validate title...
  ○ Requirement title was ignored
Validate labels...
  ○ Requirement labels was ignored
Validate users...
  ○ Requirement users was ignored
Validate teams...
  ○ Requirement teams was ignored
Approve pull request...
Enable auto-merge on pull request...
🎉  The pull request will be merged automatically!"
`;

exports[`Check run should throw an Error with pull request closed 1`] = `
"Validate context event...
Get pull request info...
❌  The pull request is not opened"
`;

exports[`Check run should throw an Error with pull request conflicts 1`] = `
"Validate context event...
Get pull request info...
Validate title...
  ○ Requirement title was ignored
Validate labels...
  ○ Requirement labels was ignored
Validate users...
  ○ Requirement users was ignored
Validate teams...
  ○ Requirement teams was ignored
Approve pull request...
❌  The pull request cannot be merged due to merge conflicts"
`;

exports[`Check run should throw an Error with pull request not mergeable 1`] = `
"Validate context event...
Get pull request info...
Validate title...
  ○ Requirement title was ignored
Validate labels...
  ○ Requirement labels was ignored
Validate users...
  ○ Requirement users was ignored
Validate teams...
  ○ Requirement teams was ignored
Approve pull request...
❌  The pull request cannot be merged"
`;

exports[`Check run should throw an Error with wrong "event" 1`] = `
"Validate context event...
❌  This action does not supports \\"lorem\\" events"
`;

exports[`Check run should throw an Error with wrong "labels" 1`] = `
"Validate context event...
Get pull request info...
Validate title...
  ○ Requirement title was ignored
Validate labels...
  ✘︎ Requirement labels is not valid
❌  The pull request does not have the required labels: lorem"
`;

exports[`Check run should throw an Error with wrong "teams" 1`] = `
"Validate context event...
Get pull request info...
Validate title...
  ○ Requirement title was ignored
Validate labels...
  ○ Requirement labels was ignored
Validate users...
  ○ Requirement users was ignored
Validate teams...
  ✘︎ Requirement teams is not valid
❌  User \\"author\\" is not part of any teams"
`;

exports[`Check run should throw an Error with wrong "title" 1`] = `
"Validate context event...
Get pull request info...
Validate title...
  ✘︎ Requirement title is not valid
❌  The title of the pull request does not match the search: lorem"
`;

exports[`Check run should throw an Error with wrong "users" 1`] = `
"Validate context event...
Get pull request info...
Validate title...
  ○ Requirement title was ignored
Validate labels...
  ○ Requirement labels was ignored
Validate users...
  ✘︎ Requirement users is not valid
❌  User \\"author\\" is not authorized"
`;
